### PR TITLE
hetzner: do not reboot instance after partitioning failed

### PR DESCRIPTION
**Pull request by danbst**
There may be errors during partitioning phase (incorrect configuration, problems with devices, etc), but they are rarely critical to reboot rescue shell.
With this change time to redeploy after partition error is reduced by 1 minute.

I still do bootstrapping phase, because nixpart version could be changed

Also I'd like to ask nixops devs:
1. Is it correct place to define new state?
2. Bootstrapping phase doesn't remove old bootstrapping tools, just overwrites them. I see, there is check on whether space is enough. So with this change redeploy after error can fail on very-very small instances (where total space in rescue tmpfs is less them 700Mb). Though it affects only "paritioning exception" scenario, so no tests would be affected

part of https://github.com/NixOS/nixops/issues/429

### CODE REVIEW

**Comment by rbvermaa** [comment]
@aszlig Do you have any opinions on this?

File: nixops/backends/hetzner.py [hunk 1]
```diff
@@ -338,12 +344,12 @@ def reboot_rescue(self, install=False, partitions=None, bootstrap=True,
...
         server.rescue.activate()
         rescue_passwd = server.rescue.password
-        if hard or (install and self.state not in (self.UP, self.RESCUE)):
+        if hard or (install and self.state not in (self.UP, self.RESCUE, self.PARTITIONING_FAILED)):
             self.log_start("sending hard reset to robot... ")
             server.reboot('hard')
         else:
             self.log_start("sending reboot command... ")
-            if self.state == self.RESCUE:
+            if self.state == self.RESCUE or self.state == self.PARTITIONING_FAILED:
```
Comment (aszlig) [inline nixops/backends/hetzner.py]: Typo: `,`

File: nixops/backends/hetzner.py [hunk 2]
```diff
@@ -591,7 +597,11 @@ def create(self, defn, check, allow_reboot, allow_recreate):
 
         if not self.vm_id:
             self.log("installing machine...")
-            self.reboot_rescue(install=True, partitions=defn.partitions)
```
Comment (aszlig) [inline nixops/backends/hetzner.py]: The same can be achieved by just checking for the RESCUE state here.

**Comment by aszlig** [comment]
I originally did this because sometimes when `blivet` has failed it could end up in an unrecoverable state, so I'm not sure whether it makes sense to reuse the rescue from that state. Maybe it would make sense to add a flag or even just emit a warning telling the user what to do if the he/she wants to throw away the current state.

**Comment by danbst** [comment]
@aszlig Thanks for review!

1 > blivet has failed it could end up in an unrecoverable state
What kind of unrecoverable state? Did rescue reboot fix it? I ask, because I've caught the "unrecoverable state" that even rescue reboot didn't fix (caused by manually configured LVM over RAID). Even `nixops destroy` didn't fix it

2 > or even just emit a warning telling the user what to do if the he/she wants to throw away the current state
I do like this approach!
```
hetznerIron> installing machine...
hetznerIron> rescue system is present and accessible
hetznerIron> partitioning...
hetznerIron> Exception: partitioning failed! Try one of these before redeploy:
  1) Check your kickstart script.
  2) Run "destroy" command to restart rescue system.
  3) Login to rescue with password XXXX and try to fix problems manually.
```

3 This isn't related to my changes, but there is an easy way to brick Nixops rescue mode. Just press Ctrl-C during rescue reboot
```
hetznerIron> installing machine...
hetznerIron> rebooting machine в??hetznerIronв?T (MY_IP) into rescue system
hetznerIron> sending reboot command... done.
hetznerIron> waiting for rescue system...Connection to MY_IP closed by remote host.
[down]....^C
$
```
Now NixOps cannot reboot instance (wrong password). I think this is because of lines https://github.com/danbst/nixops/blob/7c60f056f2438c58796c9dc5dd4bc1030bfddcdf/nixops/backends/hetzner.py#L357-L358
```
        self._wait_for_rescue(self.main_ipv4)
        self.rescue_passwd = rescue_passwd
```
I can fill issue for this if reboot on bootstrap problems will be left.

File: nixops/backends/hetzner.py (hunk 2 above)
Comment (danbst) [inline nixops/backends/hetzner.py]: I was too cautious here =)

**Comment by domenkozar** [comment]
Maybe it just could be a flag to enable this?

**Comment by danbst** [comment]
@domenkozar unfortunately, I have no free Hetzner machines to further maintain this PR.

**Pull request closed without merging**